feat(zkstack_cli): Add the ability to run the server from Docker without building it

manuelmauro opened the pull request:
## What ❔

<!-- What are the changes this PR brings about? -->
<!-- Example: This PR adds a PR template to the repo. -->
<!-- (For bigger PRs adding more context is appreciated) -->
Currently we always run the server using cargo run. If you are not a developer. it’s unnecessary to build everything from scratch.

```bash
❯ zkstack server run --mode docker

┌   ZK Stack CLI 
│
●  Starting server
│  
◆  Select image
│  ● latest 
│  ○ v25.2.0 
│  ○ v25.1.0 
│  ○ v25.0.0 
│  ○ v24.29.1 
│  ○ v24.29.0 
└
```

## Why ❔

<!-- Why are these changes done? What goal do they contribute to? What are the principles behind them? -->
<!-- Example: PR templates ensure PR reviewers, observers, and future iterators are in context about the evolution of repos. -->

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zkstack dev fmt` and `zkstack dev lint`.

manuelmauro committed:
refactor: wrap server run shell command in function

manuelmauro committed:
refactor: name arg module consistently with command module

manuelmauro committed:
refactor: use module folder for server command

manuelmauro committed:
refactor: create module for wait subcommand

manuelmauro committed:
refactor: create module for build subcommand

manuelmauro committed:
refactor: create module for run subcommand

manuelmauro committed:
refactor: create module for server run args

manuelmauro committed:
refactor: clippy

manuelmauro committed:
feat: add debug execution mode for server

manuelmauro committed:
feat: draft docker execution mode for server

manuelmauro committed:
feat: remove hardcoded volume folders

manuelmauro committed:
feat: draft option to specify docker image's tag

manuelmauro committed:
chore: update autocompletion files

manuelmauro committed:
Merge branch 'main' into manuel-run-server-from-docker

manuelmauro committed:
feat: add utils to get list of github tags

manuelmauro committed:
feat: add selector for docker image tag

manuelmauro committed:
fix: remove temporary test

manuelmauro committed:
feat: deduplicate cargo run command

manuelmauro committed:
refactor: use const for repo url

manuelmauro commented on zkstack_cli/crates/common/src/server.rs:
```diff
@@ -99,3 +107,122 @@ impl Server {
+                "docker run
+                --platform linux/amd64
+                --net=host
+                -v {configs_folder}:/configs
+                -v {chains_folder}:/chains
```
Without `/chains` we get:
```bash
Running: docker run --platform linux/amd64 --net=host -v ./configs/:/configs matterlabs/server-v2:latest --genesis-path ./chains/era/configs/genesis.yaml --wallets-path ./chains/era/configs/wallets.yaml --config-path ./chains/era/configs/general.yaml --secrets-path ./chains/era/configs/secrets.yaml --contracts-config-path ./chains/era/configs/contracts.yaml
│  

Error: failed decoding general YAML config

Caused by:
    0: failed to read ./chains/era/configs/general.yaml
    1: No such file or directory (os error 2)
```

But we may not need `/configs`.

manuelmauro committed:
refactor: remove /config volume from dockerized server

manuelmauro committed:
Merge branch 'main' into manuel-run-server-from-docker

Deniallugo commented on zkstack_cli/crates/common/src/server.rs:
```diff
@@ -99,3 +106,120 @@ impl Server {
+
+            Cmd::new(cmd!(
+                shell,
+                "docker run
+                --platform linux/amd64
+                --net=host
+                -v {chains_folder}:/chains
```
You have to have fixed paths inside container and have volumes for mounting the exact files

manuelmauro replied:
Addressed in https://github.com/matter-labs/zksync-era/pull/3314/commits/debb5070631f6669b5081cd829980f78f1ac062d

manuelmauro committed:
feat: add mode/tag to server run subcommands

manuelmauro committed:
fix: fix docker volumes/paths

manuelmauro committed:
feat: add option to run dockerized server in genesis subcommand

manuelmauro committed:
chore: regenerate autocomplete files

manuelmauro committed:
chore: update autocompletion files

manuelmauro committed:
refactor: bail instead of unwrapping

Deniallugo commented on zkstack_cli/crates/zkstack/src/commands/ecosystem/args/init.rs:
```diff
@@ -106,6 +106,8 @@ pub struct EcosystemInitArgs {
 impl EcosystemInitArgs {
     pub fn get_genesis_args(&self) -> GenesisArgs {
         GenesisArgs {
+            mode: ExecutionMode::Release,
+            tag: None,
```
is it possible to unify it as a enum?

manuelmauro replied:
Addressed in https://github.com/matter-labs/zksync-era/pull/3314/commits/f990a0ae25fe8896df91838f1f06bcf4c622c827

manuelmauro committed:
refactor: improve execution mode enum

manuelmauro committed:
refactor: clippy

manuelmauro committed:
feat: add execution mode options to missing subcommands

manuelmauro committed:
chore: regenerate autocompletion files

manuelmauro committed:
refactor: add message const for select tag prompt

Deniallugo commented on zkstack_cli/crates/common/src/server.rs:
```diff
@@ -99,3 +103,104 @@ impl Server {
+            shell,
+            "docker run
+                --platform linux/amd64
+                --net=host
+                -v {genesis_path}:/genesis.yaml
+                -v {wallets_path}:/wallets.yaml
+                -v {general_path}:/general.yaml
+                -v {secrets_path}:/secrets.yaml
```
just a nit: i'd rather put it in folder an not into the root directory

Deniallugo reviewed (commented):
LGTM, mostly nits

manuelmauro commented on zkstack_cli/crates/zkstack/src/commands/ecosystem/args/init.rs:
```diff
@@ -86,6 +86,10 @@ pub struct EcosystemInitArgs {
     /// Deploy Paymaster contract
     #[clap(long, default_missing_value = "true", num_args = 0..=1)]
     pub deploy_paymaster: Option<bool>,
+    #[arg(long, default_value = "release")]
+    pub mode: Mode,
+    #[arg(long)]
+    pub tag: Option<String>,
```
Yes it is, but I don't see a way to merge them as Clap args. Any idea?

manuelmauro committed:
refactor: wrap docker run command in a function

manuelmauro commented on zkstack_cli/crates/common/src/server.rs:
```diff
@@ -99,3 +103,104 @@ impl Server {
+            contracts_path,
+            additional_args,
+        ),
+        ExecutionMode::Docker { tag } => Cmd::new(cmd!(
```
Addressed in https://github.com/matter-labs/zksync-era/pull/3314/commits/7711a3cc869674af75496928ad68d12249c561da

manuelmauro committed:
feat: use folder for config files in docker

manuelmauro committed:
Merge branch 'main' into manuel-run-server-from-docker

manuelmauro committed:
refactor: clippy

manuelmauro commented:
Update: currently checking the problem with exposing the ports to the host machine when running the server dockerized

manuelmauro committed:
docs: improve CLI help

manuelmauro committed:
chore: update autocompletion files

manuelmauro committed:
feat: add port mapping to server's docker run

manuelmauro committed:
feat: expose all required ports from the server

Deniallugo commented on zkstack_cli/crates/zkstack/src/commands/server/run.rs:
```diff
@@ -20,16 +21,7 @@ pub(super) async fn run_server(
-
-    let rpc_port = api_config.web3_json_rpc.http_port.to_string();
-    let healthcheck_port = api_config.healthcheck.port.to_string();
+    let ports = EcosystemPortsScanner::scan(shell)?;
```
this is relevant only for docker run, should we get it  so early?

manuelmauro replied:
The problem is that we cannot get it within the `common` crate.

manuelmauro committed:
feat: adjust host to execution mode

manuelmauro committed:
fix: fix debug execution mode

manuelmauro committed:
refactor: clippy

manuelmauro committed:
Merge branch 'main' into manuel-run-server-from-docker

manuelmauro committed:
chore: update lock file

manuelmauro committed:
fix: adjust host only when required

manuelmauro committed:
fix: remove unwrapping

manuelmauro committed:
Merge branch 'main' into manuel-run-server-from-docker

manuelmauro committed:
chore: update cargo lock

manuelmauro commented:
Recent [updates to the config files on the main branch](https://github.com/matter-labs/zksync-era/commit/ff1418cbf287f40d617633eadefba2e1c31d80c3#diff-8bd4b3aaafc2fba5de99d24fe4740bd4ec6160025cd07ba31a9d936236696716L21) were not backward compatible. This causes the command `zkstack server --mode docker --tag latest` to fail with:
```bash
thread 'main' panicked at core/node/eth_sender/src/aggregator.rs:76:72:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-12-03T09:36:56.978735Z DEBUG zksync_health_check: Changed health of `prometheus_exporter` from {"status":"not_ready"} to {"status":"panicked"}
2024-12-03T09:36:56.980158Z DEBUG zksync_health_check: Changed health of `tree` from {"status":"not_ready"} to {"status":"panicked"}
│
■  Command failed to run
│  
■    Status:
│      exit status: 101
│    Stdout:
│      
│  
│    Stderr:
│      
│  
■  Failed to start server
│  
▲    0: Failed to run server
│    1: Command failed to run: docker run --platform linux/amd64 -v ./chains/era/configs/genesis.yaml:/config/genesis.yaml -v ./chains/era/configs/wallets.yaml:/config/wallets.yaml -v ./chain
```

deniallugo-ml commented:
Closing this PR as the `zkstack` CLI was substantially restructured after this PR was opened. The code paths, command structure, and configuration handling this PR targets have all changed, so the diff no longer applies to the current codebase.

If the underlying feature or fix is still needed, please open a new PR against the current CLI structure.
